content/client-lib-development-guide/features.textile

```diff
@@ -465,10 +465,23 @@ h3(#realtime-channel). Channel
-* @(RTP18)@ The realtime system reserves the right to initiate a sync of the presence members at any point once a channel is attached. A server initiated sync provides Ably with a means to send a complete list of members present on the channel at any point. The client library determines that a new sync has started whenever a @SYNC@ @ProtocolMessage@ is received with a @channel@ attribute and a new sync sequence identifier in the @channelSerial@ attribute. The @channelSerial@ is used as the sync cursor and is a two-part identifier @<sync sequence id>:<cursor value>@. If a new sequence identifier is sent from Ably, then the client library must consider that to be the start of a new sync sequence and any previous in-flight sync should be discarded. Once the cursor value is empty, then the sync operation for that sequence identifier has completed
+* @(RTP1)@ When a channel @ATTACHED@ @ProtocolMessage@ is received, the @ProtocolMessage@ may contain a @HAS_PRESENCE@ bit flag indicating that there are currently members present on the channel, see "TR3":#TR3 . If the flag is 1, the server will shortly perform a @SYNC@ operation as described in "RTP18":#RTP18 . If that flag is 0 or there is no @flags@ field, the presence map should be considered in sync immediately with no members present on the channel
+* @(RTP2)@ A @PresenceMap@ should be used to maintain a list of members present on a channel. Broadly, this is is a map of "memberKeys":#TP3h to presence messages, all with @PRESENT@ actions (during a sync there may also be ones with an @ABSENT@ action, see "RTP2f":#RTP2f ). Consult the "the Java implementation":https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/realtime/Presence.java#L578-L723 for details.
+** @(RTP2a)@ All incoming presence messages must be compared for newness with the matching member already in the @PresenceMap@, if one exists, where "matching" means they share the same @memberKey@ (or equivalently, they share both @connectionId@ and @clientId@)
+** @(RTP2b)@ To comparing for newness:
+** @(RTP2b1)@ Split the @id@ of both presence messages (which will be of the form @connid:msgSerial:index@, e.g. @aaaaaa:0:0@) on the separator @:@, to give @connectionId@, @msgSerial@, and @index@, parsing the latter two as integers (either eagerly or if/when needed)
+** @(RTP2b2)@ If the @connectionId@s are different, compare numerically by @timestamp@
```

Comment:
Example: an item in the presence map:
```
{
  clientId: 'foo',
  connectionId: 'aaaaaa',
  action: 'present',
  id: 'aaaaaa:0:0',
  timestamp: 100
}
```
A synthesized leave arrives:
```
{
  action: 14,
  id: 'xxxxxx:0',
  timestamp: 200,
  presence: [{
    clientId: 'foo',
    connectionId: 'aaaaaa',
    action: 'leave'
}]}
```
On getting this client library adds an id and timestamp to the individual presence message as usual, making it:
```
{
  id: 'xxxxxx:0:0',
  timestamp: 200
  clientId: 'foo',
  connectionId: 'aaaaaa',
  action: 'leave'
}
```
We now compare this for newness with the existing item. Both have a memberKey of `aaaaaa:foo`. So we split the id as usual, but find that the connectionId part of the ids are different, meaning we can't compare the msgSerials. So instead we compare by timestamp.

Would it be clearer if instead I said something like "If either presence message has a connectionId that is not an initial substring of its id, then compare by timestamp"? i.e. comparing the messages to themselves rather than each other. The end result is the same